### ISSUE TITLE
Define different sort order based on nested objects

### ISSUE THREAD

**Issue by ekeren** [nodeca/js-yaml#455]
In this current implementation you can't define different sort order based on nested objects

imagine I want `a' to be before `b` in root element and `b` to be before `a` under `x` element, as far as I can see,  is not possible.

I've already created a solution for this on my local branch and would love to push it. 

There are 3 options here that I see:
1. Adjust the arguments to be nested  - sort key willl be called with the key with nested name 
```javascript
test('Dumper should provide nested keys', function () {
  var nested = { a: 1, b: 2, x: { a: 2, b: 1 } };
  var expected = 'a: 1\nb: 2\nx:\n  b: 1\n  a: 2\n';
  var order = [ 'a', 'b', 'x', 'x.b', 'x.a' ];
  assert.deepEqual(yaml.safeDump(nested, {
    sortKeys: function (a, b) {
      return order.indexOf(a) - order.indexOf(b);
    }
  }), expected);
```
   - Pro: Nice clean and freindly API 
   - Con: Break backward competability 
2. Ad a third argument to sortKey - namespace (array)
```javascript
test('Dumper should provide nested keys', function () {
  var nested = { a: 1, b: 2, x: { a: 2, b: 1 } };
  var expected = 'a: 1\nb: 2\nx:\n  b: 1\n  a: 2\n';
  var rootOrder = [ 'a', 'b', 'x' ];
  var xOrder = [ 'b', 'a' ];
  assert.deepEqual(yaml.safeDump(nested, {
    sortKeys: function (a, b, namespace) {
      if (namespace && namespace.length === 1 && namespace[0] === 'x') { // an ugly way to check if namespace === ['x']
        return xOrder.indexOf(a) - xOrder.indexOf(b);
      }
      return rootOrder.indexOf(a) - rootOrder.indexOf(b);
    }
  }), expected);
```
   - Pro: brackward compatible, doens't add '.' as a special charecter 
   - Con: a sort function with a third parameter - yack! 
3. Do not fix ( i'll fork and use my own implementation) 

I think you should go with option 2, it has an API flaw but it

_Originally posted by @ekeren in https://github.com/nodeca/js-yaml/pull/188/comment#issuecomment-434783584_

**Comment by puzrin** [comment]
To be honest, now i feel something deeply wrong in this feature. Keys order in JS is deterministic. IMHO reordering is out of dumper scope and can be done in outer code.

So, my personal choice now would be (3).

**Comment by ekeren** [comment]
I see what you are saying, you mean that I need to take my json and recreate it in the desired order and then dump it using js-yaml. 

I think it doesn't make sense to have sortKey without the ability to "really" sort the keys, but in reality we got to this point. 

Maybe I'll go ahead and create a sort-json package ;)

**Comment by puzrin** [comment]
Yes. Having all in one place may look attractive, but can cause cluttered code/api, not convetient to maintain. If you have no specific requirements about highly optimized performance, it's better to keep sorter separate. That should not be difficult with lodash and so on...

I guess, existing sorter was accepted because of simplicity, and [may be] high usage probability. But increasing complexity looks "too much". I'm still not sure 100% - if anyone can explain highly-demanded use cases, i'm ready to reconsider.

**Comment by ekeren** [comment]
Hey @puzrin 

You wrote - "Keys order in JS is deterministic" I found that only for Map type the order is well defined. 

Definition of an Object from [ECMAScript Third Edition (pdf)](http://www.ecma-international.org/publications/files/ECMA-ST-ARCH/ECMA-262,%203rd%20edition,%20December%201999.pdf):

> 4.3.3 Object:
> An object is a member of the type Object. It is an unordered collection of properties each of which contains a primitive value, object, or function. A function stored in a property of an object is called a method.


Am I missing something? 

If this is indeed the case I think make sense to either support Map as input to the `dump` functions or support this behavior:

```js
test('Dumper should provide nested keys', function () {
  var nested = { a: 1, b: 2, x: { a: 2, b: 1 } };
  var expected = 'a: 1\nb: 2\nx:\n  b: 1\n  a: 2\n';
  var rootOrder = [ 'a', 'b', 'x' ];
  var xOrder = [ 'b', 'a' ];
  assert.deepEqual(yaml.safeDump(nested, {
    sortKeys: function (a, b, namespace) {
      if (namespace && namespace.length === 1 && namespace[0] === 'x') { // an ugly way to check if namespace === ['x']
        return xOrder.indexOf(a) - xOrder.indexOf(b);
      }
      return rootOrder.indexOf(a) - rootOrder.indexOf(b);
    }
  }), expected);
```

**Comment by puzrin** [comment]
Object properties order goes in order of been defined/added. You can check that in any  browser/node. I don't remember where this was said.

So, instead of sorting, you can just provide properly ordered object.

**Comment by ekeren** [comment]
thanks for the quick reply @puzrin  

I also remembered that sorting was guaranteed, but it turns out it is an implementation detail, not a language definition. Thus not guaranteed (as you can see in the above pdf)

YAML order is important (because it is human readable) so I still think this library should support some form of sorting (either by consuming Map object, or by providing the third parameter `namespace`) 

I prefer the namespace one, because:
-  sort function already exists in this library
-  it doesn't break compatibility 
-  it is already implemented in my local branch :)

**Comment by ekeren** [comment]
Although I have the ` sortKeys: function (a, b, namespace)` solution already implemented I think it makes more sense to accept `Map` instance to the `dump` function and leaving the sorting concern out of the hand of this library

**Comment by puzrin** [comment]
> I also remembered that sorting was guaranteed, but it turns out it is an implementation detail, not a language definition. Thus not guaranteed (as you can see in the above pdf)

I don't like to make changes "just for fun". It ordered objects work everywhere, why should we change anything?

**Comment by ekeren** [comment]
If it is not part of the language, there is no guarantee that it works everywhere. 

I had experience with keys enumeration change on Java from 32bit to 64bit. It wasn't fun to fix all the test just because someone leaned on implementation details. 

I have no more arguments besides what is written above. Thanks

**Comment by ekeren** [comment]
@puzrin -  should we close this?
